stacks-signer/src/signer.rs

```diff
@@ -1453,3 +1506,116 @@ impl Signer {
         Ok(())
     }
 }
+
+fn load_encrypted_signer_state<S: SignerStateStorage>(
+    storage: S,
+    id: S::IdType,
+    private_key: &Scalar,
+) -> Result<Option<v2::Signer>, PersistenceError> {
+    if let Some(encrypted_state) = storage.get_encrypted_signer_state(id)? {
+        let serialized_state = decrypt(private_key, &encrypted_state)?;
+        let state = serde_json::from_slice(&serialized_state)
+            .expect("Failed to deserialize decryoted state");
+        Ok(Some(v2::Signer::load(&state)))
+    } else {
+        Ok(None)
+    }
+}
+
+trait SignerStateStorage {
+    type IdType;
+
+    fn get_encrypted_signer_state(
+        self,
+        signer_config: Self::IdType,
+    ) -> Result<Option<Vec<u8>>, PersistenceError>;
+}
+
+impl SignerStateStorage for &mut StackerDB {
+    type IdType = SignerSlotID;
+
+    fn get_encrypted_signer_state(
+        self,
+        id: Self::IdType,
+    ) -> Result<Option<Vec<u8>>, PersistenceError> {
+        Ok(self.get_encrypted_signer_state(id)?)
+    }
+}
+
+impl SignerStateStorage for &SignerDb {
+    type IdType = u64;
+    fn get_encrypted_signer_state(
+        self,
+        id: Self::IdType,
+    ) -> Result<Option<Vec<u8>>, PersistenceError> {
+        Ok(self.get_encrypted_signer_state(id)?)
+    }
+}
+
+fn encrypt(
+    private_key: &Scalar,
+    msg: &[u8],
+    rng: &mut impl rand_core::CryptoRngCore,
+) -> Result<Vec<u8>, EncryptionError> {
+    wsts::util::encrypt(derive_encryption_key(private_key).as_bytes(), msg, rng)
+        .map_err(|_| EncryptionError::Encrypt)
+}
+
+fn decrypt(private_key: &Scalar, encrypted_msg: &[u8]) -> Result<Vec<u8>, EncryptionError> {
+    wsts::util::decrypt(derive_encryption_key(private_key).as_bytes(), encrypted_msg)
+        .map_err(|_| EncryptionError::Decrypt)
+}
+
+fn derive_encryption_key(private_key: &Scalar) -> Sha512Trunc256Sum {
+    let mut prefixed_key = "SIGNER_STATE_ENCRYPTION_KEY/".as_bytes().to_vec();
+    prefixed_key.extend_from_slice(&private_key.to_bytes());
+
+    Sha512Trunc256Sum::from_data(&prefixed_key)
+}
```

Comment:
This is the first time I've seen derived keys for encryption. I'm not knocking it at all, but what's the purpose of this? Some sort of attack where you can get the private key with a malicious payload?

Reply:
Yeah this is just for extra caution. While we could use the private key directly for encryption, this adds an extra layer in case someone manages to break the encryption as they won't be able to get the private key from this hash, since the private key is also used for other things like signing network packets.